wave_vision/src/descriptor/brisk_descriptor.cpp

```diff
@@ -0,0 +1,158 @@
+
+    // Add parameters to parser, to be loaded. If path cannot be found, throw an
+    // exception.
+    parser.addParam("radiusList", &config.radiusList);
+    parser.addParam("numberList", &config.numberList);
+    parser.addParam("dMax", &config.dMax);
+    parser.addParam("dMin", &config.dMin);
+
+    if (parser.load(config_path) != 0) {
+        throw std::invalid_argument(
+          "Failed to Load BRISKDescriptor Configuration");
+    }
+
+    // Create unused indexChange vector
+    const std::vector<int> indexChange = std::vector<int>();
+
+    // Confirm configuration is valid
+    this->checkConfiguration(config);
+
+    // Create cv::BRISK object with the desired parameters
+    this->brisk_descriptor = cv::BRISK::create(config.radiusList,
+                                               config.numberList,
+                                               config.dMax,
+                                               config.dMin,
+                                               indexChange);
+
+    // Store configuration parameters within member struct
+    this->current_config = config;
+}
+
+BRISKDescriptor::~BRISKDescriptor() {}
+
+void BRISKDescriptor::checkConfiguration(
+  const BRISKDescriptorParams &check_config) {
+    std::vector<float> rList = check_config.radiusList;
+    std::vector<int> nList = check_config.numberList;
+
+    // Check that the size of radiusList and numberList are equal and positive
+    if (rList.size() == 0) {
+        throw std::invalid_argument("No parameters in radiusList!");
+    } else if (nList.size() == 0) {
+        throw std::invalid_argument("No parameters in numberList!");
+    } else if (rList.size() != nList.size()) {
+        throw std::invalid_argument(
+          "radiusList and numberList are of unequal size!");
+    }
+
+    // Ensure all values of radiusList are positive
+    std::vector<float>::iterator rListIterator;
+
+    for (rListIterator = rList.begin(); rListIterator != rList.end();
+         rListIterator++) {
+        if (*rListIterator < 0) {
+            throw std::invalid_argument("radiusList has a negative parameter!");
+        }
+    }
+
+    // Ensure all values of numberList are positive
+    std::vector<int>::iterator nListIterator;
+
+    for (nListIterator = nList.begin(); nListIterator != nList.end();
+         nListIterator++) {
+        if (*nListIterator < 0) {
+            throw std::invalid_argument("numberList has a negative parameter!");
+        }
+    }
+
+    // Ensure dMax and dMin are both positive, and check dMax is less than dMin
+    if (check_config.dMax < 0) {
+        throw std::invalid_argument("dMax is a negative value!");
+    } else if (check_config.dMin < 0) {
+        throw std::invalid_argument("dMin is a negative value!");
+    } else if (check_config.dMax > check_config.dMin) {
+        throw std::invalid_argument("dMax is greater than dMin!");
+    }
+}
+
+BRISKDescriptorParams BRISKDescriptor::getConfiguration() {
```

Comment:
Normally you want to make these functions `const`.
Read https://isocpp.org/wiki/faq/const-correctness